### ISSUE TITLE
MSG file may leave parent task permanently pending after extracted attachment processing

### ISSUE THREAD

**Issue by kupe-dono** [pandora-analysis/pandora#966]
## Summary

While investigating a submitted Outlook MSG file (`suspicious_attachments.eml.msg`), I observed that Pandora successfully extracts the embedded attachment and creates a child task, but the original parent task remains indefinitely in "Pending Task! Data will be available soon..." The issue appears to be specific to the Outlook MSG processing path.

The same sample renamed as `.eml` reaches a final state, although it triggers a separate `observables` worker error and a preview warning.
However, when submitted as `.msg` or `.eml.msg`, the parent task consistently remains pending.

## Observed behaviour

When the MSG file is submitted:

- Pandora identifies it as an Outlook MSG/OLE file.
- The embedded attachment appears to be extracted.
- A child task is created.
- The child task is processed normally by the workers.
- The parent task never reaches a final state.
- The parent report does not contain the **Extracted** tab.
- The extractor worker unexpectedly terminates and is restarted by `WorkersManager`.

This suggests that the issue may happen after child task creation, during extracted-task relationship persistence or parent report finalization.

### 1. MSG-specific warnings are visible during processing

The submitted file follows the MSG processing path and triggers warnings related to the MSG HTML/RTF body:

```text
extract_msg.msg_classes.msg INFO:Stream "__substg1.0_10130102" was requested but could not be found. Returning `None`.

extract_msg.msg_classes.message_base INFO:HTML body was not found, attempting to generate from RTF.

File WARNING:Unable to process HTML body in MSG file ... :
'str' object has no attribute 'value'

File WARNING:Unable to generate a preview of the HTML body:
'str' object has no attribute 'value'
```

Despite these warnings, the analysis continues and the attachment extraction appears to succeed.

### 2. A child task is created and processed successfully

Example task IDs observed during one reproduction:

```text
Parent task: f0c2f960-16a3-457e-bcd1-8e2ce3190bd6
Child task:  062f99fb-640a-48eb-89ab-61c276d287a7
```

The child task is processed by multiple workers and appears to complete normally:

```text
extractor INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 0.00s
clamav INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 0.07s
observables INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 0.08s
virustotal INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 0.40s
comodo INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 2.35s
qrcode INFO:[062f99fb-640a-48eb-89ab-61c276d287a7] Runtime: 5.02s
```

However, the parent task remains pending.

### 3. The extractor worker unexpectedly terminates and is restarted

During processing of the problematic sample, the extractor worker disappears and is restarted by `WorkersManager`:

```text
2026-06-04 11:42:39,600 WorkersManager INFO:restart dead worker extractor
2026-06-04 11:42:39,601 extractor INFO:Initializing extractor-1
2026-06-04 11:42:39,617 extractor INFO:Worker is running...
```

The worker restart itself works after fixing the separate `WorkersManager` restart issue. However, the parent task still remains pending.

### 4. Full relevant log excerpt from one reproduction

```text
2026-06-04 11:42:16,417 outlookmsg INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 0.01s
2026-06-04 11:42:16,487 extract_msg.msg_classes.msg INFO:Stream "__substg1.0_10130102" was requested but could not be found. Returning `None`.
2026-06-04 11:42:16,490 extract_msg.msg_classes.message_base INFO:HTML body was not found, attempting to generate from RTF.
2026-06-04 11:42:16,488 olevba INFO:Opening OLE file /pandora/tasks/2026/06/378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2/suspicious_attachments.eml.msg
2026-06-04 11:42:16,494 olevba INFO:Check whether OLE file is PPT
2026-06-04 11:42:16,499 ole INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 0.02s
2026-06-04 11:42:16,535 File WARNING:Unable to process HTML body in MSG file e23555e8-d79f-4765-831d-b402c7a52a51: 'str' object has no attribute 'value'
2026-06-04 11:42:16,548 observables INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 0.08s
2026-06-04 11:42:16,591 unoserver INFO:Exporting to /pandora/tasks/2026/06/378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2/suspicious_attachments.eml.msg_body_txt.pdf
2026-06-04 11:42:16,613 unoserver INFO:Saved to /pandora/tasks/2026/06/378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2/suspicious_attachments.eml.msg_body_txt.pdf.
2026-06-04 11:42:16,616 extract_msg.msg_classes.msg INFO:Stream "__substg1.0_10130102" was requested but could not be found. Returning `None`.
2026-06-04 11:42:16,617 extract_msg.msg_classes.message_base INFO:HTML body was not found, attempting to generate from RTF.
2026-06-04 11:42:16,646 File WARNING:Unable to generate a preview of the HTML body: 'str' object has no attribute 'value'
2026-06-04 11:42:16,662 preview INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 0.16s
2026-06-04 11:42:16,825 virustotal INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 0.41s
2026-06-04 11:42:18,880 comodo INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 2.37s
2026-06-04 11:42:21,519 qrcode INFO:[378aa4b4-c1b9-4a00-b1a4-af1d3bc049a2] Runtime: 5.06s
2026-06-04 11:42:39,600 WorkersManager INFO:restart dead worker extractor
2026-06-04 11:42:39,601 extractor INFO:Initializing extractor-1
2026-06-04 11:42:39,617 extractor INFO:Worker is running...
```

## Reproduction

1. Submit the attached Outlook MSG file named `suspicious_attachments.eml.msg`.
2. Wait for the analysis to complete.
3. Observe that:
   - the embedded attachment is extracted;
   - a child task is created;
   - the child task appears to be processed successfully;
   - the parent task remains indefinitely pending.
4. Check the parent report.
5. Observe that the **Extracted** tab is not present.
6. Observe in the logs that the extractor worker unexpectedly terminates and is restarted.

## Possible root cause

The issue appears to happen only when Pandora follows the Outlook MSG processing path.

A likely area to investigate is the extracted-task finalization logic in the `extractor` worker.

Although a child task is created and processed successfully, the parent task:
- never exposes the **Extracted** section;
- never reaches a final state;
- appears to lose the extractor worker before parent report finalization.

This may indicate that:
- the extracted child task is not correctly associated back to the parent task;
- the extractor worker exits or crashes before persisting the extracted-task relationship;
- or the parent extractor report is never finalized after the child task processing step.

At this stage, the exact reason why the extractor worker terminates is still unknown, but the behaviour is consistently reproducible with this sample.

## Additional notes

The issue does not appear to be caused by:
- attachment extraction failure;
- corrupted MSG structure;
- worker restart failure.

The extracted attachment is valid and the generated child task is processed by workers.

The issue appears to be related to parent task finalization after MSG attachment extraction.

## Additional observation when renaming the same file to `.eml`

When the exact same MSG/OLE file is renamed to `.eml`, Pandora reaches a final state. However, this appears to follow a different processing path and triggers a separate error in the `observables` worker:

```text
observables ERROR:[f39b4551-db72-4667-99ad-7197d2600f3f] unknown error during analysis : KeyError('content_type')
Traceback (most recent call last):
  File "/pandora/pandora/workers/base.py", line 234, in run
    self.analyse(task, report, self.module == manual_worker)
  File "/pandora/pandora/workers/observables.py", line 15, in analyse
    task.init_observables_from_file()
  File "/pandora/pandora/task.py", line 299, in init_observables_from_file
    for observable_type, values in self.file.observables.items():
                                   ^^^^^^^^^^^^^^^^^^^^^
  File "/pandora/pandora/file.py", line 658, in observables
    if body['content_type'] == 'text/html':
       ~~~~^^^^^^^^^^^^^^^^
KeyError: 'content_type'
```

This may be a separate robustness issue. It suggests that some parsed email body parts may not always contain a `content_type` key.

A defensive lookup such as: body.get('content_type') could avoid crashing the `observables` worker and allow Pandora to continue processing gracefully.

The `.eml` renamed case should probably be treated as a separate issue from the MSG parent-task pending problem.

**Comment by Rafiot** [comment]
That's a lot of potential explanation. I'll need the sample file because I cannot reproduce the behavior with the ones I have. Please share it with me.

**Comment by kupe-dono** [comment]
Hi @Rafiot a colleague of mine will share the sample with you in a separated channel. :)

**Comment by Rafiot** [comment]
gotcha, it has absolutely nothing to do with anything in the issue description. 

I could reproduce it with *any* archive (or email, as it uses the extractor worker), as long as the number of replicas for the extractor module is less than `max_recurse + 1`:
* submitted file is an archive, runs the extractor worker, wait for the children to finish
* the children may or may not be archives, we check the extractor worker
* if all the extractor workers are busy, we wait for one of them to finish but none will, because they're all waiting for their children, causing the timeout.
